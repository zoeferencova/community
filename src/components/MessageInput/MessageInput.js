import React, { Component } from "react";
import styles from "../MessageLayout/MessageLayout.module.css";

export default class MessageInput extends Component {
    state = {
        message: "",
        isTyping: false,
    }

    handleSubmit = e => {
        e.preventDefault();
        this.sendMessage();
        this.setState({ message: "" })
    }

    sendMessage = () => {
        this.props.sendMessage(this.state.message)
    }

    componentWillUnmount() {
        this.stopCheckingTyping()
    }

    sendTyping = () => {
        this.lastUpdateTime = Date.now()
        if (!this.state.isTyping) {
            this.setState({ isTyping: true })
            this.props.sendTyping(true)
            this.startCheckingTyping()
        }
    }

    startCheckingTyping = () => {
        this.typingInterval = setInterval(() => {
            if ((Date.now() - this.lastUpdateTime) > 300) {
                this.setState({ isTyping: false })
                this.stopCheckingTyping()
            }
        }, 300)
    }

    stopCheckingTyping = () => {
        if (this.typingInterval) {
            clearInterval(this.typingInterval)
            this.props.sendTyping(false)
        }
    }
    
    render() {
        const { message } = this.state;
        return (
            <div className={styles.messageInput}>
                <form onSubmit={this.handleSubmit} className={styles.messageForm}>
                    <input 
                        id="message" 
                        ref={"messageinput"} 
                        type="text" 
                        className={styles.formControl} 
                        value={message} 
                        autoComplete={"off"} 
                        placeholder = "Type a message..." 
                        // onKeyUp={e => {e.keyCode !== 13 && this.sendTyping()}} 
                        onChange={({target}) => {this.setState({ message: target.value })}}></input>
                    <button disabled={message.length < 1} type="submit" className={styles.send}><img className={styles.sendIcon} src="https://img.icons8.com/ios-glyphs/96/89a1fc/paper-plane.png"/></button>
                </form>
            </div>
        )
    }
}