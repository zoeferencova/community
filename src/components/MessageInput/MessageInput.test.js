import React from "react";
import { shallow } from "enzyme";
import MessageInput from "./MessageInput";

describe("MessageInput", () => {
    let wrapper, mockProps;

    beforeAll(() => {
        mockProps = {
            sendMessage: jest.fn()
        }

        wrapper = shallow(<MessageInput { ...mockProps } />)
    })

    it("renders MessageInput component", () => {        
        expect(wrapper).toMatchSnapshot()
    });

    it("calls sendMessage function and clears message value when message input form is submitted", () => {
        wrapper.setState({ message: "test" })
        wrapper.find(".messageForm").simulate("submit")

        expect(mockProps.sendMessage).toHaveBeenCalledTimes(1)
        expect(wrapper.find("#message").prop("value")).toEqual("");
    });
});