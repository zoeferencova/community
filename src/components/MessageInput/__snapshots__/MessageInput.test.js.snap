// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageInput renders MessageInput component 1`] = `
<div
  className="messageInput"
>
  <form
    className="messageForm"
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="formControl"
      id="message"
      onChange={[Function]}
      placeholder="Type a message..."
      type="text"
      value=""
    />
    <button
      className="send"
      disabled={true}
      type="submit"
    >
      <img
        alt="send icon"
        className="sendIcon"
        src="https://img.icons8.com/ios-glyphs/96/89a1fc/paper-plane.png"
      />
    </button>
  </form>
</div>
`;
